streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -186,8 +187,24 @@ private void startRestoration(final Map<TopicPartition, StateRestorer> initializ
                 restorer.setStartingOffset(restoreConsumer.position(restorer.partition()));
                 restorer.restoreStarted();
             } else {
-                restoreConsumer.seekToBeginning(Collections.singletonList(restorer.partition()));
-                needsPositionUpdate.add(restorer);
+                final StreamTask task = active.restoringTaskFor(restorer.partition());
+
+                // If checkpoint does not exist it means the task was not shutdown gracefully before;
+                // and in this case if EOS is turned on we should wipe out the state and re-initialize the task
+                if (task.isEosEnabled()) {
+                    log.info("No checkpoint found for task {} state store {} changelog {} with EOS turned on. " +
+                            "Reinitializing the task.", task.id, restorer.storeName(), restorer.partition());
+                    task.reinitializeStateStoresForPartitions(Collections.singleton(restorer.partition()));
+                    restoreConsumer.seekToBeginning(Collections.singleton(restorer.partition()));
+
+                } else {
+                    log.info("Restoring task {}'s state store {} from beginning of the changelog {} ", task.id, restorer.storeName(), restorer.partition());
+
+                    restoreConsumer.seekToBeginning(Collections.singletonList(restorer.partition()));
```

Comment:
We `seekToBeginning` in both cases -- should we extract and share this?

Also, even for non-EOS, what would be the disadvantage in wiping out the store, too? We could still have some writes to the store that might not be cleaned up, and it seems we can get this cleanup for free?